### ISSUE TITLE
Task values are unbounded

### ISSUE THREAD

**Issue by SabreCat** [HabitRPG/habitica#3251]
The fantastic [HabitRPG Unofficial User Data Display](https://oldgods.net/habitrpg/habitrpg_user_data_display.html#) revealed to me that task values just keep cranking up, or down, over time. This took me by surprise, since the comments in this code block suggest they're supposed to be capped:

```
# Min/max on task redness
currVal =
if task.value < -47.27 then -47.27
else if task.value > 21.27 then 21.27
else task.value
```

Perhaps I was mistaken to take it at face value, but "Min/max on task redness" suggests that tasks are supposed to keep getting redder until they hit -47.27, then stop; or keep getting bluer until they hit 21.27, then stop. But a well-aged To-Do or utterly mastered Habit can have absolute values in the hundreds or even thousands.

On one level, that's all well and good; finishing that To-Do you've been putting off for months should be super valuable, cool 'nuff. But as soon as we introduce things like class abilities into the equation, things get wonky. With abilities acting on the task magnitude, doing something unintuitive like Brutal Smashing an ancient To-Do (which, per the ability's ordinary use, makes no sense) can let you deal hundreds of damage to a boss monster.

Moreover, the drop system assumes that task values range from -47.27 to 21.27; absent those bounds, really blue tasks will "wrap around" and start becoming more likely to drop items, rather than running down to a minimum 2% base chance and stopping there.

**Comment by Alys** [comment]
It might not be as serious as you fear (although be a bit cautious about what I write here since I'm relatively new at this).

Firstly, the -47.27 / 21.27 limit IS being applied (i.e., that code isn't being ignored because of a bug), it's just not applied in a permanent way to each task - and from the look of the code, it's not meant to be, although the code might be wrong. :)

Instead, those limits ARE applied but only temporarily when certain calculations are being made. Here's when the -47.27 / 21.27 limits DO take effect:
- when working out how much to increase/decrease the value of a task
- when increasing the "user.stat.training" attributes (for automatic allocation, I think)
- when calculating damage to a quest boss when a daily or todo is ticked off
- when determining the XP and GP you gain from ticking off a task
- when determining the HP you loose from not doing a daily (thank goodness!!)

For drop chances... I'm still uncertain. I would want to study the randomDrop function a bit more before I could make an objective statement. However from a purely subjective point of view: Since you (it was you wasn't it??) re-did the drops algorithm, I've been very happy with how it's working for me. I don't get too many drops because of the drop cap, and they ARE spread out very evenly over the day (yay!!!), and they "feel" random. Often I don't even reach the drop cap, so I certainly don't have too high a drop chance. Hence if the -47.27/21.27 issue IS affecting drop chances, it's not doing so in a way that is noticably harmful, from my subjective point of view.

As for skills...

[Brutal Smash](http://habitrpg.wikia.com/wiki/Warrior_Skills) affects task value but is not AFFECTED BY it (i.e., a very blue/red task does NOT let you do massive damage to a boss). The biggest problem with this skill and the -47.27/21.27 issue is that if you have a VERY red daily, then the skill won't have any noticable impact on that daily's value. And that probably IS a problem, but if the -47.27/21.27 issue isn't changed, then perhaps the skill could be improved by making it adjust the value of a task by a PROPORTION of its current value. (That's a random idea I just thought of; it might be a terrible one.)

[Burst of Flames](http://habitrpg.wikia.com/wiki/Mage_Skills) has a similar problem with adding to the task value (i.e., a very small relative effect on very red/blue tasks). For XP gain, you can get 75 XP by using it just once on a very blue daily, and you can do 50 damage to a boss from one cast on the same daily. However that's perfectly okay because I am a mage. ;) ;) A high-powered mage who's saved up a large amount of mana can take down the largest boss we have in a single day (although from memory, they have to also do many tasks that day to get additional mana). This is really good for those of us who like to do solo quests. I suspect though that you might find it unsuitable from any other perpective. :)

[Pickpocket](http://habitrpg.wikia.com/wiki/Rogue_Skills) is affected by task value but is capped at 25 GP, so it's not a concern.

[Backstab](http://habitrpg.wikia.com/wiki/Rogue_Skills) is insanely effective on very blue dailies. INSANELY! I think I got 4000GP once and similar XP, from memory. Insane. I don't use that spell when I'm playing a rogue because it's just too much. However, I believe it's being modified, and if not, a simple modification could be a cap, such as the one Pickpocket has.

The other skills don't use task value (unless I've missed some).

So in summary, I think most things are okay with the -47.27/21.27 cap working as it does now.
Brutal Smash is perhaps a bit UNDER-powered.
Backstab is definitely overpowered, but being fixed (or would be easy to fix).
Burst of Flames is perfect just as it is. Perfect. ;)

(_sad sigh_ I'm going to loose my wonderfully over-powered Burst of Flames skill, aren't I?)

And now, if you notice anything wrong with my analysis, I would LOVE to know about it, either from a technical/code point of view or from a how-we-want-the-game-to-behave point of view.

Thanks for the "fantastic" comment. :)

**Comment by SabreCat** [comment]
OK, that is indeed reassuring. Though it's confusing in the code and leads to mistakes like mine in the drop chance algorithm:

```
chance = _.min([Math.abs(task.value - 21.27),37.5])/150+.02   # Base drop chance is a percentage based on task value. Typical fresh task: 15%. Very ripe task: 25%. Very blue task: 2%.
```

The "21.27" is in there because I thought that was the ceiling for blue tasks, leading to mastered tasks dwindling down until they hit 0/150+.02 = 2% base drop chance. Instead, what now happens is that they dip to a 2% chance at value 21.27, then proceed to climb back up until maxing out at 27% just like an increasingly red task.

That said, I realize a little better now why we don't put hard caps on these; graphs hitting a ceiling would be unfortunate, especially for Challenge judging. But the drop chance algorithm is a little weird and unintuitive at the moment as a result of this.

**Comment by davidjiang-gorsline** [comment]
Personally - I use Habitica to help me identify which habits I need to focus more on based on task value/colour. It is quite helpful but it becomes frustrating when a long neglected habit doesn't change colour after a good amount of streaks. I checked the task value of the task and it's -200+, which will take a lot to get it back to normal. Similar rationale for good habits that has been maintaining for a good amount of time it becomes invincible. It would be great if we can cap it so the task value and its corresponding dailies colour can be an indication on how strong the habit is and like a solid 10 day streak can turn a maroon habit to a yellow one and missing 10 days in a roll can turn a teal habit into a weaker colour. Just for some the team's consideration :)
